src/ripple/app/tx/impl/SetOracle.cpp

```diff
@@ -0,0 +1,250 @@
+{
+    if (!ctx.rules.enabled(featurePriceOracle))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    auto const& dataSeries = ctx.tx.getFieldArray(sfPriceDataSeries);
+    if (dataSeries.size() == 0 || dataSeries.size() > maxOracleDataSeries)
+        return temBAD_ARRAY_SIZE;
+
+    auto invalidLength = [&](auto const& sField, std::size_t length) {
+        return ctx.tx.isFieldPresent(sField) &&
+            ctx.tx[sField].length() > length;
+    };
+
+    if (invalidLength(sfProvider, maxOracleProvider) ||
+        invalidLength(sfURI, maxOracleURI) ||
+        invalidLength(sfAssetClass, maxOracleSymbolClass))
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+SetOracle::preclaim(PreclaimContext const& ctx)
+{
+    auto const sleSetter =
+        ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount)));
+    if (!sleSetter)
+        return terNO_ACCOUNT;
+
+    // lastUpdateTime must be within 30 seconds of the last closed ledger
+    using namespace std::chrono;
+    std::size_t const closeTime =
+        duration_cast<seconds>(ctx.view.info().closeTime.time_since_epoch())
+            .count();
+    std::size_t const lastUpdateTime = ctx.tx[sfLastUpdateTime];
+    if (lastUpdateTime < closeTime ||
+        lastUpdateTime > (closeTime + maxLastUpdateTimeDelta))
+        return tecINVALID_UPDATE_TIME;
+
+    hash_set<uint256> pairs;
+    for (auto const& entry : ctx.tx.getFieldArray(sfPriceDataSeries))
+    {
+        if (!entry.isFieldPresent(sfAssetPrice))
+            return temMALFORMED;
+        auto const hash = tokenPairHash(entry);
+        if (pairs.contains(hash))
+            return tecDUPLICATE;
+        pairs.emplace(hash);
+    }
+
+    if (auto const sle = ctx.view.read(keylet::oracle(
+            ctx.tx.getAccountID(sfAccount), ctx.tx[sfOracleSequence])))
+    {
+        // update
+
+        // lastUpdateTime must be more recent than the previous one
+        if (ctx.tx[sfLastUpdateTime] <= (*sle)[sfLastUpdateTime])
+            return tecINVALID_UPDATE_TIME;
+
+        if (ctx.tx[sfAccount] != sle->getAccountID(sfOwner))
+            return tecNO_PERMISSION;
+
+        if (ctx.tx.isFieldPresent(sfProvider) ||
+            ctx.tx.isFieldPresent(sfAssetClass))
+            return temMALFORMED;
+
+        for (auto const& entry : sle->getFieldArray(sfPriceDataSeries))
+        {
+            auto const hash = tokenPairHash(entry);
+            if (!pairs.contains(hash))
+                pairs.emplace(hash);
+        }
+    }
+    else
+    {
+        // create
+
+        if (!ctx.tx.isFieldPresent(sfProvider) ||
+            !ctx.tx.isFieldPresent(sfAssetClass))
+            return temMALFORMED;
+    }
+
+    if (pairs.size() > maxOracleDataSeries)
+        return temBAD_ARRAY_SIZE;
+
+    auto const add = pairs.size() > 5 ? 2 : 1;
```

Comment:
Is it supposed to be `>` or `>=`?

Reply:
the code is correct as-is. here's the spec PDF (second bullet point)
![image](https://github.com/XRPLF/rippled/assets/21219765/13e8e6b3-84f2-49b2-b1c2-362b6d6b1e66)